# Titanic: Machine Learning from Disaster

# Definição do Escopo

* Realizar o processo de Data Munging no dataset do tytanic;
* Gerar arquivos CSVs para o próximos projeto para aplicar os algoritmos machine learning;
* Os arquivos devem passar pelo processo de:
    * Análise Exploratória de Dados;
    * Análise Estatística de Dados;
    * Pré-Processamento;


# Objetivo

Gerar dataset(csv) limpo e formatado para aplicar no projeto algoritmos de ML para prever: 
* Que tipos de pessoas provavelmente sobreviveriam?

# Funções

In [1]:
def print_full(x):
    pd.set_option('display.max_rows', len(x))
    print(x)
    pd.reset_option('display.max_rows')

# Coleta de dados

https://www.kaggle.com/c/titanic/data

<b>PassengerId</b> = Passageiro ID<br>
<b>Survived</b> = Sobreviventes<br>
<b>Pclass</b> = Classe do Ingresso<br>
<b>Name</b> = Nome<br>
<b>Sex</b> = Sexo<br>
<b>Age</b> = Idade<br>
<b>SibSp</b> = Parentes Irmão/Irmã/Marido/Esposa<br>
<b>Parch</b> = Parentes Mãe/Pai/Filhos/Enteado<br>
<b>Ticket</b> = Passagem Número<br>
<b>Fare</b> = Passagem Preço<br>
<b>Cabin</b> = Cabine<br>
<b>Embarked</b> = Porto de Embarque(C = Cherbourg; Q = Queenstown; S = Southampton)<br>

## Fontes

In [2]:
### Importando as libs
import pandas as pd
import numpy as np

In [3]:
### Coletando as fontes de dados
train_csv = pd.read_csv('train.csv', sep = ',', encoding = 'UTF-8')
test_csv = pd.read_csv('test.csv', sep = ',', encoding = 'UTF-8')

## Validação - Importação dos dados

In [4]:
test_csv.head(2)

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S

In [5]:
train_csv.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

## Renomeando - Colunas

In [6]:
### Colunas de Teste
test=test_csv.copy()
colunasTeste=['PassageiroId','Classe','Nome','Sexo','Idade','ParentesIrmao','ParentesFilhos',
              'PassagemNumero','PassagemPreco','Cabine','PortoEmbarque']
test.columns=colunasTeste

In [7]:
### Colunas de Treino
train=train_csv.copy()
colunasTreino = colunasTeste[:]
colunasTreino.insert(1,'Sobreviventes')
train.columns=colunasTreino

## Validação - Renomeação das colunas

In [8]:
test.head(2)

PassageiroId  Classe                              Nome    Sexo  Idade  \
0           892       3                  Kelly, Mr. James    male   34.5   
1           893       3  Wilkes, Mrs. James (Ellen Needs)  female   47.0   

   ParentesIrmao  ParentesFilhos PassagemNumero  PassagemPreco Cabine  \
0              0               0         330911         7.8292    NaN   
1              1               0         363272         7.0000    NaN   

  PortoEmbarque  
0             Q  
1             S

In [9]:
train.head(2)

PassageiroId  Sobreviventes  Classe  \
0             1              0       3   
1             2              1       1   

                                                Nome    Sexo  Idade  \
0                            Braund, Mr. Owen Harris    male   22.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38.0   

   ParentesIrmao  ParentesFilhos PassagemNumero  PassagemPreco Cabine  \
0              1               0      A/5 21171         7.2500    NaN   
1              1               0       PC 17599        71.2833    C85   

  PortoEmbarque  
0             S  
1             C

## Join Metadados

In [10]:
### União do Treino e Teste 
frames = [train[colunasTeste], test]
join = pd.concat(frames)

## Coleta de dados - Conclusão

<b>Dados de treino:</b> Possuí informações dos sobreviventes, então podemos utilizar algoritmos machine learning supervisionado.
<br>
<b>Dados de teste:</b> As previsões dos dados de teste deve ser enviado para o Kaggle, conforme o modelo: gender_submission.csv
<br>
<b>Join Metadados:</b> A união dos datasets de treino e teste vai ajudar no processo de análise exploratória dos dados.
<br>
<b>Join Metadados:</b> Foi excluído a coluna sobreviventes.

# Análise Exploratória de Dados

## Visualizar - 1ª Linhas

In [11]:
### Treino
train.head(20)

PassageiroId  Sobreviventes  Classe  \
0              1              0       3   
1              2              1       1   
2              3              1       3   
3              4              1       1   
4              5              0       3   
5              6              0       3   
6              7              0       1   
7              8              0       3   
8              9              1       3   
9             10              1       2   
10            11              1       3   
11            12              1       1   
12            13              0       3   
13            14              0       3   
14            15              0       3   
15            16              1       2   
16            17              0       3   
17            18              1       2   
18            19              0       3   
19            20              1       3   

                                                 Nome    Sexo  Idade  \
0                             Braund, Mr. Owen Harris    male   22.0   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38.0   
2                              Heikkinen, Miss. Laina  female   26.0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35.0   
4                            Allen, Mr. William Henry    male   35.0   
5                                    Moran, Mr. James    male    NaN   
6                             McCarthy, Mr. Timothy J    male   54.0   
7                      Palsson, Master. Gosta Leonard    male    2.0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   27.0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female   14.0   
10                    Sandstrom, Miss. Marguerite Rut  female    4.0   
11                           Bonnell, Miss. Elizabeth  female   58.0   
12                     Saundercock, Mr. William Henry    male   20.0   
13                        Andersson, Mr. Anders Johan    male   39.0   
14               Vestrom, Miss. Hulda Amanda Adolfina  female   14.0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female   55.0   
16                               Rice, Master. Eugene    male    2.0   
17                       Williams, Mr. Charles Eugene    male    NaN   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female   31.0   
19                            Masselmani, Mrs. Fatima  female    NaN   

    ParentesIrmao  ParentesFilhos    PassagemNumero  PassagemPreco Cabine  \
0               1               0         A/5 21171         7.2500    NaN   
1               1               0          PC 17599        71.2833    C85   
2               0               0  STON/O2. 3101282         7.9250    NaN   
3               1               0            113803        53.1000   C123   
4               0               0            373450         8.0500    NaN   
5               0               0            330877         8.4583    NaN   
6               0               0             17463        51.8625    E46   
7               3               1            349909        21.0750    NaN   
8               0               2            347742        11.1333    NaN   
9               1               0            237736        30.0708    NaN   
10              1               1           PP 9549        16.7000     G6   
11              0               0            113783        26.5500   C103   
12              0               0         A/5. 2151         8.0500    NaN   
13              1               5            347082        31.2750    NaN   
14              0               0            350406         7.8542    NaN   
15              0               0            248706        16.0000    NaN   
16              4               1            382652        29.1250    NaN   
17              0               0            244373        13.0000    NaN   
18              1               0            345763        18.0000    NaN   
19              0               0              

In [12]:
### Teste
test.head(20)

PassageiroId  Classe                                               Nome  \
0            892       3                                   Kelly, Mr. James   
1            893       3                   Wilkes, Mrs. James (Ellen Needs)   
2            894       2                          Myles, Mr. Thomas Francis   
3            895       3                                   Wirz, Mr. Albert   
4            896       3       Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
5            897       3                         Svensson, Mr. Johan Cervin   
6            898       3                               Connolly, Miss. Kate   
7            899       2                       Caldwell, Mr. Albert Francis   
8            900       3          Abrahim, Mrs. Joseph (Sophie Halaut Easu)   
9            901       3                            Davies, Mr. John Samuel   
10           902       3                                   Ilieff, Mr. Ylio   
11           903       1                         Jones, Mr. Charles Cresson   
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
13           905       2                               Howard, Mr. Benjamin   
14           906       1  Chaffee, Mrs. Herbert Fuller (Carrie Constance...   
15           907       2      del Carlo, Mrs. Sebastiano (Argenia Genovesi)   
16           908       2                                  Keane, Mr. Daniel   
17           909       3                                  Assaf, Mr. Gerios   
18           910       3                       Ilmakangas, Miss. Ida Livija   
19           911       3              Assaf Khalil, Mrs. Mariana (Miriam")"   

      Sexo  Idade  ParentesIrmao  ParentesFilhos    PassagemNumero  \
0     male   34.5              0               0            330911   
1   female   47.0              1               0            363272   
2     male   62.0              0               0            240276   
3     male   27.0              0               0            315154   
4   female   22.0              1               1           3101298   
5     male   14.0              0               0              7538   
6   female   30.0              0               0            330972   
7     male   26.0              1               1            248738   
8   female   18.0              0               0              2657   
9     male   21.0              2               0         A/4 48871   
10    male    NaN              0               0            349220   
11    male   46.0              0               0               694   
12  female   23.0              1               0             21228   
13    male   63.0              1               0             24065   
14  female   47.0              1               0       W.E.P. 5734   
15  female   24.0              1               0     SC/PARIS 2167   
16    male   35.0              0               0            233734   
17    male   21.0              0               0              2692   
18  female   27.0              1               0  STON/O2. 3101270   
19  female   45.0              0               0              2696   

    PassagemPreco Cabine PortoEmbarque  
0          7.8292    NaN             Q  
1          7.0000    NaN             S  
2          9.6875    NaN             Q  
3          8.6625    NaN             S  
4         12.2875    NaN             S  
5          9.2250    NaN             S  
6          7.6292    NaN             Q  
7         29.0000    NaN             S  
8          7.2292    NaN             C  
9         24.1500    NaN             S  
10         7.8958    NaN             S  
11        26.0000    NaN             S  
12        82.2667    B45             S  
13        26.0000    NaN             S  
14        61.1750    E31             S  
15        27.7208    NaN             C  
16        12.3500    NaN             Q  
17         7.2250    NaN             C  
18         7.9250    NaN             S  
19         7.2250    NaN             C

## Quais as dimensões do dataset?


In [13]:
### Treino
train.shape

(891, 12)

In [14]:
### Teste
test.shape

(418, 11)

In [15]:
### JOIN
join.shape

(1309, 11)

<b>Dados de treino:</b> A quantidade de dados de treino aparentemente não é suficiente para ter um modelo com alta precisão.
<br>

## Quais são os tipos das colunas?

In [16]:
### Treino
train.dtypes

PassageiroId        int64
Sobreviventes       int64
Classe              int64
Nome               object
Sexo               object
Idade             float64
ParentesIrmao       int64
ParentesFilhos      int64
PassagemNumero     object
PassagemPreco     float64
Cabine             object
PortoEmbarque      object
dtype: object

In [17]:
### Teste
test.dtypes

PassageiroId        int64
Classe              int64
Nome               object
Sexo               object
Idade             float64
ParentesIrmao       int64
ParentesFilhos      int64
PassagemNumero     object
PassagemPreco     float64
Cabine             object
PortoEmbarque      object
dtype: object

In [18]:
### Join
join.dtypes

PassageiroId        int64
Classe              int64
Nome               object
Sexo               object
Idade             float64
ParentesIrmao       int64
ParentesFilhos      int64
PassagemNumero     object
PassagemPreco     float64
Cabine             object
PortoEmbarque      object
dtype: object

## Quais as colunas que tem valores nulos?

In [19]:
### Treino
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassageiroId      891 non-null int64
Sobreviventes     891 non-null int64
Classe            891 non-null int64
Nome              891 non-null object
Sexo              891 non-null object
Idade             714 non-null float64
ParentesIrmao     891 non-null int64
ParentesFilhos    891 non-null int64
PassagemNumero    891 non-null object
PassagemPreco     891 non-null float64
Cabine            204 non-null object
PortoEmbarque     889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [20]:
### Teste
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassageiroId      418 non-null int64
Classe            418 non-null int64
Nome              418 non-null object
Sexo              418 non-null object
Idade             332 non-null float64
ParentesIrmao     418 non-null int64
ParentesFilhos    418 non-null int64
PassagemNumero    418 non-null object
PassagemPreco     417 non-null float64
Cabine            91 non-null object
PortoEmbarque     418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [21]:
### Join
join.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1309 entries, 0 to 417
Data columns (total 11 columns):
PassageiroId      1309 non-null int64
Classe            1309 non-null int64
Nome              1309 non-null object
Sexo              1309 non-null object
Idade             1046 non-null float64
ParentesIrmao     1309 non-null int64
ParentesFilhos    1309 non-null int64
PassagemNumero    1309 non-null object
PassagemPreco     1308 non-null float64
Cabine            295 non-null object
PortoEmbarque     1307 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 122.7+ KB


## Quais total de valores nulos por coluna?

In [22]:
### Treino
train.isnull().sum()

PassageiroId        0
Sobreviventes       0
Classe              0
Nome                0
Sexo                0
Idade             177
ParentesIrmao       0
ParentesFilhos      0
PassagemNumero      0
PassagemPreco       0
Cabine            687
PortoEmbarque       2
dtype: int64

In [23]:
### Teste
test.isnull().sum()

PassageiroId        0
Classe              0
Nome                0
Sexo                0
Idade              86
ParentesIrmao       0
ParentesFilhos      0
PassagemNumero      0
PassagemPreco       1
Cabine            327
PortoEmbarque       0
dtype: int64

In [24]:
### Join
join.isnull().sum()

PassageiroId         0
Classe               0
Nome                 0
Sexo                 0
Idade              263
ParentesIrmao        0
ParentesFilhos       0
PassagemNumero       0
PassagemPreco        1
Cabine            1014
PortoEmbarque        2
dtype: int64

## Sumário Estatístico

Resumo estatístico do DataFrame, com quartis, mediana, etc

<ul>
<li>count = quantidade de registros não nulo</li>
<li>mean = média aritmética</li>
<li>std = desvio padrão</li>
<li>min = minímo valor</li>
<li>25% = Primeiro Quartil</li>
<li>50% = Segundo Quartil = Mediana</li>
<li>75% = Terceiro Quartil</li>
<li>max = maior valor</li>
</ul>

### Dados - Brutos

In [25]:
### Treino 
train.describe()

PassageiroId  Sobreviventes      Classe       Idade  ParentesIrmao  \
count    891.000000     891.000000  891.000000  714.000000     891.000000   
mean     446.000000       0.383838    2.308642   29.699118       0.523008   
std      257.353842       0.486592    0.836071   14.526497       1.102743   
min        1.000000       0.000000    1.000000    0.420000       0.000000   
25%      223.500000       0.000000    2.000000   20.125000       0.000000   
50%      446.000000       0.000000    3.000000   28.000000       0.000000   
75%      668.500000       1.000000    3.000000   38.000000       1.000000   
max      891.000000       1.000000    3.000000   80.000000       8.000000   

       ParentesFilhos  PassagemPreco  
count      891.000000     891.000000  
mean         0.381594      32.204208  
std          0.806057      49.693429  
min          0.000000       0.000000  
25%          0.000000       7.910400  
50%          0.000000      14.454200  
75%          0.000000      31.000000  
max          6.000000     512.329200

In [26]:
### Teste
test.describe()

PassageiroId      Classe       Idade  ParentesIrmao  ParentesFilhos  \
count    418.000000  418.000000  332.000000     418.000000      418.000000   
mean    1100.500000    2.265550   30.272590       0.447368        0.392344   
std      120.810458    0.841838   14.181209       0.896760        0.981429   
min      892.000000    1.000000    0.170000       0.000000        0.000000   
25%      996.250000    1.000000   21.000000       0.000000        0.000000   
50%     1100.500000    3.000000   27.000000       0.000000        0.000000   
75%     1204.750000    3.000000   39.000000       1.000000        0.000000   
max     1309.000000    3.000000   76.000000       8.000000        9.000000   

       PassagemPreco  
count     417.000000  
mean       35.627188  
std        55.907576  
min         0.000000  
25%         7.895800  
50%        14.454200  
75%        31.500000  
max       512.329200

In [27]:
### Join
join.describe()

PassageiroId       Classe        Idade  ParentesIrmao  ParentesFilhos  \
count   1309.000000  1309.000000  1046.000000    1309.000000     1309.000000   
mean     655.000000     2.294882    29.881138       0.498854        0.385027   
std      378.020061     0.837836    14.413493       1.041658        0.865560   
min        1.000000     1.000000     0.170000       0.000000        0.000000   
25%      328.000000     2.000000    21.000000       0.000000        0.000000   
50%      655.000000     3.000000    28.000000       0.000000        0.000000   
75%      982.000000     3.000000    39.000000       1.000000        0.000000   
max     1309.000000     3.000000    80.000000       8.000000        9.000000   

       PassagemPreco  
count    1308.000000  
mean       33.295479  
std        51.758668  
min         0.000000  
25%         7.895800  
50%        14.454200  
75%        31.275000  
max       512.329200

### Sumário Estatístico(Dados Brutos) - Conclusão 

* **PassageiroId:** 
    * Treino: Desconsiderar os valores relacionados a coluna.
    * Teste:  Desconsiderar os valores relacionados a coluna.
    * União:  Desconsiderar os valores relacionados a coluna.
* **Sobreviventes:** 
    * Treino: Aproximadamente 38% sobreviveu - mean 0.383838
* **Classe:** 
    * A mediana é a melhor representação da Classe.(50% do Quartil);  
    * Corresponde a 3ª classe como maioria nos dados de treino e test.
    * A mediana é a melhor representação da Classe;     
* **Idade:** 
* **ParentesIrmao:**
* **ParentesFilhos:** 
* **PassagemPreco:**

### É possível trabalhar com dataset removendo os valores nulos?

In [28]:
join.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 270 entries, 1 to 414
Data columns (total 11 columns):
PassageiroId      270 non-null int64
Classe            270 non-null int64
Nome              270 non-null object
Sexo              270 non-null object
Idade             270 non-null float64
ParentesIrmao     270 non-null int64
ParentesFilhos    270 non-null int64
PassagemNumero    270 non-null object
PassagemPreco     270 non-null float64
Cabine            270 non-null object
PortoEmbarque     270 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 25.3+ KB


#### Conclusão
* Se remover todas as colunas com valores nulos fica inviável trabalhar com dataset;
* A coluna Cabine existe vários valores nulos;


### Dados - Sem nulos

In [29]:
colunasSemCabine = colunasTeste[:]
colunasSemCabine.remove('Cabine')

In [30]:
#Valor total - bruto: 891
#Valor total - Sem Nulos: 712
train_dropna=train.dropna(subset=colunasSemCabine)
train_dropna.describe()

PassageiroId  Sobreviventes      Classe       Idade  ParentesIrmao  \
count    712.000000     712.000000  712.000000  712.000000     712.000000   
mean     448.589888       0.404494    2.240169   29.642093       0.514045   
std      258.683191       0.491139    0.836854   14.492933       0.930692   
min        1.000000       0.000000    1.000000    0.420000       0.000000   
25%      222.750000       0.000000    1.000000   20.000000       0.000000   
50%      445.000000       0.000000    2.000000   28.000000       0.000000   
75%      677.250000       1.000000    3.000000   38.000000       1.000000   
max      891.000000       1.000000    3.000000   80.000000       5.000000   

       ParentesFilhos  PassagemPreco  
count      712.000000     712.000000  
mean         0.432584      34.567251  
std          0.854181      52.938648  
min          0.000000       0.000000  
25%          0.000000       8.050000  
50%          0.000000      15.645850  
75%          1.000000      33.000000  
max          6.000000     512.329200

In [31]:
#Valor total - bruto: 418
#Valor total - Sem Nulos: 331
test_dropna=test.dropna(subset=colunasSemCabine)
test_dropna.describe()

PassageiroId      Classe       Idade  ParentesIrmao  ParentesFilhos  \
count    331.000000  331.000000  331.000000     331.000000      331.000000   
mean    1100.232628    2.141994   30.181269       0.483384        0.398792   
std      122.910180    0.846251   14.104573       0.875004        0.811582   
min      892.000000    1.000000    0.170000       0.000000        0.000000   
25%      992.500000    1.000000   21.000000       0.000000        0.000000   
50%     1100.000000    2.000000   27.000000       0.000000        0.000000   
75%     1210.500000    3.000000   39.000000       1.000000        1.000000   
max     1307.000000    3.000000   76.000000       8.000000        6.000000   

       PassagemPreco  
count     331.000000  
mean       40.982087  
std        61.228558  
min         0.000000  
25%         8.050000  
50%        16.000000  
75%        40.633350  
max       512.329200

In [32]:
#Valor total - bruto: 1309
#Valor total - Sem Nulos: 1043
join_dropna=join.dropna(subset=colunasSemCabine)
join_dropna.describe()

PassageiroId       Classe        Idade  ParentesIrmao  ParentesFilhos  \
count   1043.000000  1043.000000  1043.000000    1043.000000     1043.000000   
mean     655.391179     2.209012    29.813202       0.504314        0.421860   
std      377.527036     0.840685    14.366254       0.913080        0.840655   
min        1.000000     1.000000     0.170000       0.000000        0.000000   
25%      326.500000     1.000000    21.000000       0.000000        0.000000   
50%      662.000000     2.000000    28.000000       0.000000        0.000000   
75%      973.500000     3.000000    39.000000       1.000000        1.000000   
max     1307.000000     3.000000    80.000000       8.000000        6.000000   

       PassagemPreco  
count    1043.000000  
mean       36.603024  
std        55.753648  
min         0.000000  
25%         8.050000  
50%        15.750000  
75%        35.077100  
max       512.329200

## Balanceamento

### Qual a quantidade de sobreviventes nos dados de treino?

In [33]:
train.groupby('Sobreviventes').size()

Sobreviventes
0    549
1    342
dtype: int64

### Qual a quantidade de sobreviventes nos dados de treino sem os campos nulos?

In [34]:
train_dropna.groupby('Sobreviventes').size()

Sobreviventes
0    424
1    288
dtype: int64

### Existe algum porto de embarque que sobreviveu mais pessoas ou todos sobreviveram por não embarcar?

In [35]:
### Resposta: Não
train.groupby(['Sobreviventes','PortoEmbarque']).size()

Sobreviventes  PortoEmbarque
0              C                 75
               Q                 47
               S                427
1              C                 93
               Q                 30
               S                217
dtype: int64

### Quem tinha filhos sobreviveu mais do que morreu?

In [36]:
### Resposta: Não. Mas que não tinha filhos morreu mais e ninguem sobreviveu que tinha 4 e 6 parentes.
### 4 e 6 = 0 = Morreu
### 0,1,2,3,5 = 1 = Talvez
train.groupby(['Sobreviventes','ParentesFilhos','Sexo','Classe']).size()

Sobreviventes  ParentesFilhos  Sexo    Classe
0              0               female  1           1
                                       2           5
                                       3          35
                               male    1          63
                                       2          81
                                       3         260
               1               female  2           1
                                       3          13
                               male    1          10
                                       2           7
                                       3          22
               2               female  1           2
                                       3          17
                               male    1           3
                                       2           3
                                       3          15
               3               female  3           1
                               male    3           1


### Existem nos dados teste pessoas que tem mais que 6 filhos?

In [37]:
### Sim 9. 
test.groupby(['ParentesFilhos']).size()

ParentesFilhos
0    324
1     52
2     33
3      3
4      2
5      1
6      1
9      2
dtype: int64

### Quem tinhas mais parentes sobreviveu?

In [38]:
### Resposta: Não. Mas que não tinha parantes morreu mais e ninguem sobreviveu que tinha 5,6,7,8 parentes.
### 5,6,7,8   = 0 = Morreu
### 0,1,2.3,4   = 1 = Talvez
train.groupby(['Sobreviventes','ParentesIrmao']).size()

Sobreviventes  ParentesIrmao
0              0                398
               1                 97
               2                 15
               3                 12
               4                 15
               5                  5
               8                  7
1              0                210
               1                112
               2                 13
               3                  4
               4                  3
dtype: int64

### Existem nos dados teste pessoas que tem mais que 8 parentes?

In [39]:
### Não.
test.groupby(['ParentesIrmao']).size()

ParentesIrmao
0    283
1    110
2     14
3      4
4      4
5      1
8      2
dtype: int64

### Sobreviveu mais mulheres ou homens?

In [40]:
### Resposta: Sobreviveu mais mulheres e morreu menos mulheres.
train.groupby(['Sobreviventes','Sexo']).size()

Sobreviventes  Sexo  
0              female     81
               male      468
1              female    233
               male      109
dtype: int64

### Existem pessoas com mesmo número de passagem?

In [41]:
### Resposta: Sim
mesmo_numero_passagem=train.groupby(['Sobreviventes','PassagemNumero']).size()
mesmo_numero_passagem

Sobreviventes  PassagemNumero   
0              110413               1
               110465               2
               111240               1
               111320               1
               112050               1
               112052               1
               112058               1
               112059               1
               112379               1
               113028               1
               113043               1
               113050               1
               113051               1
               113056               1
               113059               1
               113501               1
               113503               1
               113509               1
               113510               1
               113514               1
               113767               1
               113773               1
               113776               1
               113781               2
               113784               1
               11

### As pessoas com mesmo número de passagem sobreviveram?

In [42]:
### Resposta: Não
df_mesmo_numero_passagem = pd.DataFrame({'Sobreviventes':mesmo_numero_passagem.index.get_level_values(0),'Passagem':mesmo_numero_passagem.index.get_level_values(1), 'PassagemQuantidade':mesmo_numero_passagem.values})
df_mesmo_numero_passagem.groupby(['Sobreviventes','PassagemQuantidade']).size()

Sobreviventes  PassagemQuantidade
0              1                     428
               2                      29
               3                       3
               4                       3
               5                       2
               6                       3
               7                       2
1              1                     195
               2                      52
               3                      10
               4                       2
               5                       1
dtype: int64

### Existem pessoas na mesma cabine?

In [43]:
### Resposta: Sim
mesma_cabine=train.groupby(['Sobreviventes','Cabine']).size()
mesma_cabine

Sobreviventes  Cabine     
0              A10            1
               A14            1
               A19            1
               A24            1
               A32            1
               A36            1
               A5             1
               A7             1
               B102           1
               B19            1
               B22            1
               B30            1
               B37            1
               B38            1
               B51 B53 B55    1
               B58 B60        1
               B71            1
               B82 B84        1
               B86            1
               B94            1
               C110           1
               C111           1
               C118           1
               C123           1
               C124           2
               C128           1
               C2             1
               C22 C26        2
               C23 C25 C27    2
               C30            1
             

### As pessoas na mesma cabine sobreviveram?

In [44]:
### Resposta: Não
df_mesma_cabine = pd.DataFrame({'Sobreviventes':mesma_cabine.index.get_level_values(0),'Cabine':mesma_cabine.index.get_level_values(1), 'CabineQuantidade':mesma_cabine.values})
df_mesma_cabine.groupby(['Sobreviventes','CabineQuantidade']).size()

Sobreviventes  CabineQuantidade
0              1                   56
               2                    6
1              1                   70
               2                   28
               3                    2
               4                    1
dtype: int64

### Qual a idade das pessoas que sobreviveram e morreram?

In [45]:
print_full(train.groupby(['Sobreviventes','Idade']).size())

Sobreviventes  Idade
0              1.00      2
               2.00      7
               3.00      1
               4.00      3
               6.00      1
               7.00      2
               8.00      2
               9.00      6
               10.00     2
               11.00     3
               14.00     3
               14.50     1
               15.00     1
               16.00    11
               17.00     7
               18.00    17
               19.00    16
               20.00    12
               20.50     1
               21.00    19
               22.00    16
               23.00    10
               23.50     1
               24.00    15
               24.50     1
               25.00    17
               26.00    12
               27.00     7
               28.00    18
               28.50     2
               29.00    12
               30.00    15
               30.50     2
               31.00     9
               32.00     9
               32.50     1
       

### Qual a menor/maior idade que um homem e mulher morreu/sobreviveu?

In [46]:
### Respostas:
### Menor Mulher Morreu: 2 anos
### Menor Mulher Sobreviveu: 7 meses

### Menor Homem Morreu: 1 ano
### Menor Homem Sobreviveu: 4 meses


### Maior Mulher Morreu: 57 anos
### Maior Mulher Sobreviveu: 63 anos

### Maior Homem Morreu: 74 anos
### Maior Homem Sobreviveu: 80 anos
print_full(train.groupby(['Sobreviventes','Idade','Sexo']).size())

Sobreviventes  Idade  Sexo  
0              1.00   male       2
               2.00   female     4
                      male       3
               3.00   female     1
               4.00   male       3
               6.00   female     1
               7.00   male       2
               8.00   female     1
                      male       1
               9.00   female     4
                      male       2
               10.00  female     1
                      male       1
               11.00  female     1
                      male       2
               14.00  female     1
                      male       2
               14.50  female     1
               15.00  male       1
               16.00  female     1
                      male      10
               17.00  female     1
                      male       6
               18.00  female     5
                      male      12
               19.00  male      16
               20.00  female     2
                      male

### Qual a menor/maior preço de passagem de quem morreu/sobreviveu pagou?

In [51]:
### Respostas:
### Sobrevieveu maior: 512
### Sobrevieveu menor: 0
### Morreu maior menor: 263 
### Morreu maior maior: 0 
print_full(train.groupby(['Sobreviventes','PassagemPreco']).size())

Sobreviventes  PassagemPreco
0              0.0000           14
               4.0125            1
               5.0000            1
               6.2375            1
               6.4375            1
               6.4500            1
               6.4958            2
               6.7500            2
               6.8583            1
               6.9500            1
               6.9750            1
               7.0458            1
               7.0500            7
               7.0542            2
               7.1250            4
               7.2250            9
               7.2292           11
               7.2500           12
               7.3125            1
               7.4958            2
               7.5208            1
               7.5500            3
               7.6292            1
               7.6500            3
               7.7250            1
               7.7292            1
               7.7333            2
               7.7375     

### Existe uma classe que sobreviveu ou morreu todos?

In [55]:
### Resposta: Não
train.groupby(['Sobreviventes','Classe']).size()

Sobreviventes  Classe
0              1          80
               2          97
               3         372
1              1         136
               2          87
               3         119
dtype: int64

### Existe alguma classe que só homem ou mulher sobreviveram/morreram?

In [57]:
### Resposta: Não
train.groupby(['Sobreviventes','Classe','Sexo']).size()

Sobreviventes  Classe  Sexo  
0              1       female      3
                       male       77
               2       female      6
                       male       91
               3       female     72
                       male      300
1              1       female     91
                       male       45
               2       female     70
                       male       17
               3       female     72
                       male       47
dtype: int64

### Existem nomes duplicados?

In [59]:
### Resposta: Não
print_full(train.groupby(['Nome']).size())

Nome
Abbing, Mr. Anthony                                                                   1
Abbott, Mr. Rossmore Edward                                                           1
Abbott, Mrs. Stanton (Rosa Hunt)                                                      1
Abelson, Mr. Samuel                                                                   1
Abelson, Mrs. Samuel (Hannah Wizosky)                                                 1
Adahl, Mr. Mauritz Nils Martin                                                        1
Adams, Mr. John                                                                       1
Ahlin, Mrs. Johan (Johanna Persdotter Larsson)                                        1
Aks, Mrs. Sam (Leah Rosen)                                                            1
Albimona, Mr. Nassef Cassem                                                           1
Alexander, Mr. William                                                                1
Alhomaki, Mr. Ilmari Rudolf